common/src/dns.rs

```diff
@@ -0,0 +1,108 @@
+}
+
+impl SeedFeature {
+    pub fn as_code(&self) -> &'static str {
+        match self {
+            Self::NODE_NETWORK => &"x1",
+            Self::NODE_BLOOM => &"x5",
+            Self::NODE_WITNESS => &"x9",
+            Self::FULL_NODE => &"xd",
+        }
+    }
+}
+
+/// Seeds implementation
+pub struct Seeds<'a> {
+    /// The network to work load the dns
+    network: Network,
+    /// The default port of the node
+    port: u16,
+    /// List of feature supported by nakamoto
+    features: &'a [SeedFeature],
+}
+
+impl<'a> Seeds<'a> {
+    /// Create a new seed with the requirements
+    pub fn new(network: Network, port: u16, features: &'a [SeedFeature]) -> Self {
+        Seeds {
+            network,
+            port,
+            features,
+        }
+    }
+
+    pub fn load(&self) -> Vec<SocketAddr> {
+        let seeds = self.from_network();
+        self.filter_by_feature(seeds)
+    }
+
+    fn from_network(&self) -> &[&str] {
+        match self.network {
+            Network::Mainnet => &[
+                "seed.bitcoin.sipa.be",          // Pieter Wuille
+                "dnsseed.bluematt.me",           // Matt Corallo
+                "dnsseed.bitcoin.dashjr.org",    // Luke Dashjr
+                "seed.bitcoinstats.com",         // Christian Decker
+                "seed.bitcoin.jonasschnelli.ch", // Jonas Schnelli
+                "seed.btc.petertodd.org",        // Peter Todd
+                "seed.bitcoin.sprovoost.nl",     // Sjors Provoost
+                "dnsseed.emzy.de",               // Stephan Oeste
+                "seed.bitcoin.wiz.biz",          // Jason Maurice
+                "seed.cloudhead.io",             // Alexis Sellier
+            ],
+            Network::Testnet => &[
+                "testnet-seed.bitcoin.jonasschnelli.ch",
+                "seed.tbtc.petertodd.org",
+                "seed.testnet.bitcoin.sprovoost.nl",
+                "testnet-seed.bluematt.me",
+            ],
+            Network::Regtest => &[], // No seeds
+        }
+    }
+
+    fn filter_by_feature(&self, seed_list: &[&str]) -> Vec<SocketAddr> {
+        let mut seeds_addr = Vec::new();
+        for feature in self.features {
+            let feature_code = feature.as_code();
+            debug!("Filtering by feature {}", feature_code);
+
+            for seed in seed_list {
+                let sub_seed = format!("{}.{}:{}", feature_code, seed, self.port);
+                debug!("Sub seed {} checking ...", sub_seed);
+                if let Ok(hosts) = (*sub_seed).to_socket_addrs() {
+                    for host in hosts {
+                        seeds_addr.push(host);
+                    }
+                } else {
+                    // we admit the failure at this point because not all the seed
+                    // support all the feature.
+                    debug!("Seed {} doesn't support feature {}", seed, feature_code);
+                }
+            }
+        }
+        seeds_addr
+    }
+}
```

Comment:
For PRs I tend to prefer as minimal a diff as possible, as it makes it easier for me to review. Eg. I don't want to have to re-review my own code. So I'd suggest not moving existing code around.

Reply:
Thanks for clarifying this @cloudhead, in other words, do you want that I put the check of sub-seed in the `nakamoto_comm::network`?

I usually maintain the code struct of the repository, and I'm sorry if I move your code out of your class, but I did that because I think have a DNS seed module makes easy future change.

BTW, if you want I can revert the change

Reply:
Thanks for clarifying this @cloudhead, in other words, do you want that I put the check of sub-seed in the `nakamoto_comm::network`?

I usually maintain the code struct of the repository, and I'm sorry if I move your code out of your class, but I did that because I think have a DNS seed module makes easy future change.

BTW, if you want I can revert the change

Reply:
> Thanks for clarifying this @cloudhead, in other words, do you want that I put the check of sub-seed in the `nakamoto_comm::network`?
> 

Yes, whatever results in the smallest diff.

> I usually maintain the code struct of the repository, and I'm sorry if I move your code out of your class, but I did that because I think have a DNS seed module makes easy future change.
> 
> BTW, if you want I can revert the change

Yeah, that would be better for now. It makes my job easier to review code when less of it changes :)
It's okay if doing that means that some code is not in the optimal place. This is something that can be changed later on.

Generally, I try to not mix code refactors with change in functionality, as you have to then review two things at once.

Reply:
>Generally, I try to not mix code refactors with change in functionality, as you have to then review two things at once.

I agree with your point of view, thanks. I will change this PR, or close it and open a new one with the change in the `nakamoto_comm::network`, thanks